src/post-processors/exchange-rates/exchange-rates.ts

```diff
@@ -0,0 +1,58 @@
+import { ExchangeRate } from '../../model'
+import { Block, Context } from '../../processor'
+import { useProcessorState } from '../../utils/state'
+import { Currency } from './currencies'
+import { getPrice } from './price-routing'
+
+const useExchangeRates = () =>
+  useProcessorState('exchange-rates', new Map<string, ExchangeRate>())
+
+export const process = async (ctx: Context) => {
+  const [rates] = useExchangeRates()
+  if (rates.size > 0) {
+    ctx.log.info({ count: rates.size }, 'exchange-rates')
+    await ctx.store.insert([...rates.values()])
+  }
+}
+
+export const ensureExchangeRate = async (
+  ctx: Context,
+  block: Block,
+  base: Currency,
+  quote: Currency,
+) => {
+  const [exchangeRates] = useExchangeRates()
+  const pair = `${base}_${quote}`
+  const blockNumber = block.header.height
+  const id = `${blockNumber}:${pair}`
+  let exchangeRate = exchangeRates.get(id)
+  if (exchangeRate) return
+
+  const timestamp = new Date(block.header.timestamp)
+  const price = await getPrice(ctx, block, base, quote).catch((err) => {
+    ctx.log.info({ base, quote, err })
+    throw err
+  })
+  if (price) {
+    exchangeRate = new ExchangeRate({
+      id,
+      timestamp,
+      blockNumber,
+      pair,
+      base,
+      quote,
+      rate: price,
+    })
+    exchangeRates.set(id, exchangeRate)
+  }
+}
+
+export const ensureExchangeRates = async (
+  ctx: Context,
+  block: Block,
+  pairs: [Currency, Currency][],
+) => {
+  await Promise.all(
+    pairs.map(([base, quote]) => ensureExchangeRate(ctx, block, base, quote)),
```

Comment:
nit: Can probably do some filtering here. If the event was for tokens `frxETH` and `stETH`, could probably just query those tokens' price on-chain. For the rest, can find the previous value in DB and use that. Will reduce the number of RPC calls, while ensuring we have a price for all assets at a given block.

But probably better to leave this as it is for now rather than complicating it. If we run into rate-limit issues or performance bottlenecks, we could consider doing it then

Reply:
Problem is if we did that then we would be showing some of the values at wrong prices for the block numbers shown. For any block number where we have a value for a currency in our database we are going to need the exchange rate for it at the same block.